### PR TITLE
Threading API

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/fiber/FiberNodes.java [hunk 1]
```diff
@@ -145,7 +145,7 @@ public static void start(RubyContext context, ThreadManager threadManager, Dynam
         threadManager.initializeValuesBasedOnCurrentJavaThread(Layouts.FIBER.getRubyThread(fiber), pThreadID);
 
         Layouts.THREAD.getFiberManager(Layouts.FIBER.getRubyThread(fiber)).registerFiber(fiber);
-        context.getSafepointManager().enterThread();
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/fiber/FiberNodes.java]: The place to enterThread() and leaveThread() was carefully selected (e.g., the thread has a `pThreadID` before entering and knows its Java thread), so this change might introduce problems as they are now done around the rest of the initialization.

File: src/test/java/org/truffleruby/MiscTest.java [hunk 2]
```diff
@@ -37,4 +42,92 @@ public void testMembersAndStringUnboxing() {
...
+        }
+    }
+
+    @Ignore("Ruby doesn't shut down fibres properly")
+    @Test
+    public void testFiberShutdown() {
+        try (Context context = Context.newBuilder().allowCreateThread(true).build()) {
+            context.eval("ruby", "[1, 2, 3].each.tap { |e| e.next }.tap { |e| e.next }");
+        }
+    }
```
Comment (eregon) [inline src/test/java/org/truffleruby/MiscTest.java]: What happened to those tests?

File: src/main/java/org/truffleruby/core/fiber/FiberNodes.java (hunk 1 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/core/fiber/FiberNodes.java]: We didn't merge this PR.

File: src/test/java/org/truffleruby/MiscTest.java (hunk 2 above)
Comment (chrisseaton) [reply src/test/java/org/truffleruby/MiscTest.java]: We didn't merge this PR.

**Pull request closed without merging**